Cabal/Distribution/Simple/Register.hs

```diff
@@ -411,9 +411,31 @@ generalInstalledPackageInfo adjustRelIncDirs pkg abi_hash lib lbi clbi installDi
     IPI.importDirs         = [ libdir installDirs | hasModules ],
     -- Note. the libsubdir and datasubdir templates have already been expanded
     -- into libdir and datadir.
+    --
+    -- TODO: Right now, only GHC > 8.0.1 supports a package database entry with
+    -- a field that indicates where the dynamic libraries are installed. When
+    -- more compilers start to support this, we should drop the
+    -- `supportsLibraryDynDir` check here, and just handle `IPI.libraryDynDirs`
+    -- in the actual package registration function belonging to those particular
+    -- compilers.
     IPI.libraryDirs        = if hasLibrary
-                               then libdir installDirs : extraLibDirs bi
-                               else                      extraLibDirs bi,
+                               then if supportsLibraryDynDir
+                                       then libdir installDirs : extraLibDirs bi
+                                       -- If the compiler doesn't understand
+                                       -- a separate entry for dynamic
+                                       -- library directories, we add the
+                                       -- directory to the list of "normal"
+                                       -- library directories
+                                       else libdir installDirs :
+                                            dynlibdir installDirs :
+                                            extraLibDirs bi
+                               else extraLibDirs bi,
+    IPI.libraryDynDirs     = if supportsLibraryDynDir
+                               then if hasLibrary
+                                      then dynlibdir installDirs :
+                                           extraLibDirs bi
+                                      else extraLibDirs bi
+                               else [],
```

Comment:
I think it might be easier to pull these out to the where clause (like the absinc, relink) and share the logic, e.g.
```
(libdir, dynlibdir)
  | not hasLibrary
  = (extraLibDirs bi, [])
    -- the dynamic-library-dirs defaults to the library-dirs if not specified,
    -- so this works if the dynamic-library-dirs field is supported or not

  | supportsLibraryDynDir comp
  = (libdir    installDirs : extraLibDirs bi,
     dynlibdir installDirs : extraLibDirs bi)

  | otherwise
  = (libdir installDirs : dynlibdir installDirs : extraLibDirs bi, [])
    -- the compiler doesn't understand the dynamic-library-dirs field so we
    -- add the dyn directory to the "normal" list in the library-dirs field
```
but do double check I've transcribed the logic correctly! :-)